File: COCI '16 Contest 4 #3 Kas.py
# https://dmoj.ca/problem/coci16c4p3
# Dynamic programming, optimized version

MN = 200002
n = int(input())
arr = [int(input()) for _ in range(n)]
total = sum(arr)
inf = 1 << 30

prev_dp = [-inf] * MN
prev_dp[0] = 0  # base case: have nothing

for val in arr:
    dp = [0] * MN
    for i in range(total + 1):
        dp[i] = max(prev_dp[i],
                    prev_dp[abs(i - val)] + val,  # abs() since dp states are symmetric
                    prev_dp[i + val] + val)
    prev_dp = dp

print(total - prev_dp[0] // 2)

File: COCI '16 Contest 4 #3 Kas.cpp
// https://dmoj.ca/problem/coci16c4p3
// almost same as https://leetcode.com/problems/tallest-billboard
// brute force memoization
// check Python version for smarter code

#include <bits/stdc++.h>
using namespace std;

const int MM = 200002;
const int offset = 100001;
const int MN = 501;
const int inf = 1 << 30;

int dp[501][MM];
int arr[MN];
int N;

int solve(int idx, int cur) {
    if (idx == N) {
        if (cur == 0) return 0;  // valid split
        else return -inf;
    }
    if (dp[idx][cur + offset] != -1)  // cache
        return dp[idx][cur + offset];

    int res = max({solve(idx + 1, cur),
                   solve(idx + 1, cur + arr[idx]) + arr[idx],
                   solve(idx + 1, cur - arr[idx]) + arr[idx]});
    return dp[idx][cur + offset] = res;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N;
    int total = 0;
    for (int i = 0; i < N; i++) {
        cin >> arr[i];
        total += arr[i];
    }

    memset(dp, -1, sizeof(dp));
    int both = solve(0, 0) / 2;

    int diff = total - both * 2;
    cout << both + diff << "\n";

    return 0;
}
